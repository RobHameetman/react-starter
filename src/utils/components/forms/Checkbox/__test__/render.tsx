import { ReactElement } from 'react';
import { screen, fireEvent, render as _render } from '@testing-library/react';

const getResults = () =>
	screen.queryAllByRole('checkbox') as ReadonlyArray<HTMLElement>;

const getComponent = (index = 0) => getResults()[index];

const getInput = ($component = getComponent()) =>
	$component.querySelector('input') as HTMLInputElement;

const getLabel = ($component = getComponent()) =>
	$component.querySelector('label') as HTMLLabelElement;

const getText = ($label = getLabel()) => $label.innerHTML;

const getContainer = ($component = getComponent()) =>
	$component.querySelector('[id$="-container"]') as HTMLDivElement;

const getBox = ($component = getComponent()) =>
	$component.querySelector('[id$="-box"]') as HTMLDivElement;

const getCheckmark = ($component = getComponent()) =>
	$component.querySelector('[id$="-checkmark"]') as HTMLDivElement;

const getDashmark = ($component = getComponent()) =>
	$component.querySelector('[id$="-dashmark"]') as HTMLDivElement;

const getFill = ($component = getComponent()) =>
	$component.querySelector('[id$="-fill"]') as HTMLDivElement;

const getStrikethrough = ($component = getComponent()) =>
	$component.querySelector('[id$="-strikethrough"]') as HTMLDivElement;

const getChecked = ($input = getInput()) => $input.checked;

export const render = (Component: ReactElement, index = 0) => {
	const result = _render(Component);

	const $component = getComponent(index);
	const $input = getInput($component);
	const $label = getLabel($component);
	const $text = getText($label);
	const $container = getContainer($component);
	const $box = getBox($component);
	const $checkmark = getCheckmark($component);
	const $dashmark = getDashmark($component);
	const $fill = getFill($component);
	const $strikethrough = getStrikethrough($component);
	const checked = getChecked($input);
	const initiallyChecked = checked;

	const pressCheckbox = () => {
		fireEvent.pointerDown($component, { pointerType: 'mouse' });

		return getChecked();
	};

	return {
		$box,
		$checkmark,
		$component,
		$container,
		$dashmark,
		$fill,
		$input,
		$label,
		$root: $component,
		$strikethrough,
		$text,
		initiallyChecked,
		pressCheckbox,
		result,
	};
};
