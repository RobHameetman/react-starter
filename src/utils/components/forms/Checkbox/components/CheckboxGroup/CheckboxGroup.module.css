.checkboxGroup {
	display: flex;
	flex-direction: row;
	margin-bottom: 0.3125em;

	&:global(.vertical) {
		flex-direction: column;
		width: fit-content;
	}

	.label {
		color: var(--primary-color);
		font-size: 0.875rem;
		font-weight: var(--font-weight-semibold);

		&.xs {
			font-size: 0.625rem;
		}

		&.sm {
			font-size: 0.75rem;
		}

		&.lg {
			font-size: 1rem;
		}

		&.xl {
			font-size: 1.125rem;
		}
	}
}
