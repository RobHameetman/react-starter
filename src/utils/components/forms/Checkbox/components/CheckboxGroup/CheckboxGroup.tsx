import { $FC } from 'react';
import { prefersReducedMotion } from '@app/utils/functions/accessibility/prefersReducedMotion';
import { isFunction } from '@app/utils/functions/check/js/core/isFunction';
import { isString } from '@app/utils/functions/check/js/core/isString';
import { cssClasses } from '@app/utils/functions/misc/cssClasses';
import { noop } from '@app/utils/functions/misc/noop';
import { uniqueId } from '@app/utils/functions/misc/uniqueId';
import { withId } from '@app/utils/hocs/withId';
import { usePropsWithChildren } from '@app/utils/hooks/react/usePropsWithChildren';
import { useSemanticAsProp } from '@app/utils/hooks/react/useSemanticAsProp';
import type { Animatable } from '@app/utils/types/props/Animatable';
import type { Changeable } from '@app/utils/types/props/Changeable';
import type { Disablable } from '@app/utils/types/props/Disablable';
import type { Identifiable } from '@app/utils/types/props/Identifiable';
import type { Intentable } from '@app/utils/types/props/Intentable';
import type { Labelable } from '@app/utils/types/props/Labelable';
import type { Nameable } from '@app/utils/types/props/Nameable';
import type { Polymorphic } from '@app/utils/types/props/Polymorphic';
import type { Sizable } from '@app/utils/types/props/Sizable';
import type { Stylable } from '@app/utils/types/props/Stylable';
import type { Valuable } from '@app/utils/types/props/Valuable';
import { handlerPropsOf } from '@app/utils/types/react/HandlerPropsOf';
import { propsWithoutHandlers } from '@app/utils/types/react/PropsWithoutHandlers';
import { CheckboxProvider } from '../../contexts/CheckboxProvider';
import type { CheckboxProps } from '../../Checkbox';
import styles from './CheckboxGroup.module.css';

/**
 * Compositional prop types for the {@link CheckboxGroup} component.
 */
type ComposedProps = Pick<
	CheckboxProps,
	'checked' | 'label' | 'readonly' | 'strikethrough'
> &
	Animatable &
	Changeable &
	Disablable &
	Identifiable &
	Intentable &
	Labelable &
	Nameable &
	Polymorphic &
	Sizable &
	Stylable &
	Valuable<ReadonlyArray<string>>;

/**
 * Prop types for the {@link CheckboxGroup} component.
 */
export interface CheckboxGroupProps extends ComposedProps {
	/**
	 * [Optional] Set the delimiter for the checkbox group values.
	 * @defaultValue - `','`
	 */
	readonly delimiter?: string;

	/**
	 * [Optional] If `true`, sets the styling of the checkbox group to be vertical.
	 * @defaultValue - `false`
	 */
	readonly vertical?: boolean;
}

/**
 * @TODO - A short description of the component here.
 */
export const CheckboxGroup: $FC<CheckboxGroupProps> = withId(
	({
		animated = !prefersReducedMotion(),
		as: _as = 'fieldset',
		checked,
		className = '',
		children: _children,
		delimiter = ',',
		disabled,
		id = uniqueId('checkboxGroup'),
		intent,
		label = '',
		name = id,
		readonly,
		size,
		strikethrough,
		value,
		vertical = false,
		...remainingProps
	}) => {
		const As = useSemanticAsProp({ as: _as });

		const css = cssClasses(
			'checkboxGroup',
			styles.checkboxGroup,
			{ vertical },
			className,
		);

		const labelCss = cssClasses(styles.label, styles[size || 'md']);

		const eventHandlers = handlerPropsOf(remainingProps);
		const props = propsWithoutHandlers(remainingProps);

		const children = usePropsWithChildren({
			children: _children,
			criteria: {
				checked: (childProps) => {
					const childValue =
						childProps.value ||
						(isString(childProps.children)
							? childProps.children
							: childProps.name);

					return checked || value?.includes(childValue as string) || null;
				},
			},
			props: {
				animated,
				checked,
				disabled,
				intent,
				readonly,
				size,
				strikethrough,
				...eventHandlers,
			},
		});

		return (
			<CheckboxProvider delimiter={delimiter} value={value}>
				<As className={css} id={id} name={name} role="group" {...props}>
					<legend id={`${id}-label`} className={labelCss}>
						{label}
					</legend>
					{children}
				</As>
			</CheckboxProvider>
		);
	},
);
