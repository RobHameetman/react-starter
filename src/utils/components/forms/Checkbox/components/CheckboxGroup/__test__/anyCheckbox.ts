export const anyCheckbox = expect.stringMatching(/Test 1|Test 2|Test 3/);
