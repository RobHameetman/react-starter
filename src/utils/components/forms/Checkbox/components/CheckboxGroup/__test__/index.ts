export * from './anyCheckbox';
export * from './render';
