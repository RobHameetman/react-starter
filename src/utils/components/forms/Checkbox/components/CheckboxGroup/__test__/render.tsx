import { ReactElement } from 'react';
import { faker } from '@faker-js/faker';
import { fireEvent, screen, render as _render } from '@testing-library/react';

const getResults = () =>
	screen.queryAllByRole('group') as ReadonlyArray<HTMLElement>;

const getComponent = (index = 0) => getResults()[index];

const getLabel = ($component = getComponent()) =>
	$component.querySelector('legend') as HTMLLegendElement;

const getChildren = ($component = getComponent()) =>
	Array.from<HTMLDivElement>($component.querySelectorAll('[role="checkbox"]'));

const getChildCount = ($children = getChildren()) => $children.length;

const getInputs = ($component = getComponent()) =>
	Array.from($component.querySelectorAll('input'));

const getLabels = ($component = getComponent()) =>
	Array.from($component.querySelectorAll('label'));

const getText = ($labels = getLabels()) =>
	$labels.map(({ innerHTML }) => innerHTML);

const getContainers = ($component = getComponent()) =>
	Array.from<HTMLDivElement>($component.querySelectorAll('[id$="-container"]'));

const getCheckmarks = ($component = getComponent()) =>
	Array.from<HTMLDivElement>($component.querySelectorAll('[id$="-checkmark"]'));

const getDashmarks = ($component = getComponent()) =>
	Array.from<HTMLDivElement>($component.querySelectorAll('[id$="-dashmark"]'));

const getStrikethroughs = ($component = getComponent()) =>
	Array.from<HTMLDivElement>(
		$component.querySelectorAll('[id$="-strikethrough"]'),
	);

const getBoxes = ($component = getComponent()) =>
	Array.from<HTMLDivElement>($component.querySelectorAll('[id$="-box"]'));

const getFills = ($component = getComponent()) =>
	Array.from<HTMLDivElement>($component.querySelectorAll('[id$="-fill"]'));

const getStates = ($inputs = getInputs()) =>
	$inputs.map(({ checked }) => checked);

const getChecked = (states = getStates()) =>
	states.filter((checked) => checked).length;

const getUnchecked = (states = getStates()) =>
	states.filter((checked) => !checked).length;

const getValues = ($inputs = getInputs()) => $inputs.map(({ value }) => value);
const getValue = (values = getValues()) => values[0];

export const render = (Component: ReactElement, index = 0) => {
	const result = _render(Component);

	const $component = getComponent(index);
	const $label = getLabel($component);
	const $children = getChildren($component);
	const childCount = getChildCount($children);
	const $boxes = getBoxes($component);
	const $containers = getContainers($component);
	const $checkmarks = getCheckmarks($component);
	const $dashmarks = getDashmarks($component);
	const $strikethroughs = getStrikethroughs($component);
	const $fills = getFills($component);
	const $inputs = getInputs($component);
	const $labels = getLabels($component);
	const $text = getText($labels);
	const states = getStates($inputs);
	const checked = getChecked(states);
	const unchecked = getUnchecked(states);
	const values = getValues($inputs);
	const value = getValue(values);

	const pressCheckbox = (value: string) => {
		$children.forEach(($child) => {
			const $text = $child.querySelector('label')?.innerHTML;

			if ($text === value) {
				fireEvent.pointerDown($child, { pointerType: 'mouse' });
			}
		});

		return getChecked();
	};

	const pressCheckboxes = (...values: ReadonlyArray<string>) => {
		$children.forEach(($child) => {
			const $text = $child.querySelector('label')?.innerHTML;

			if (values.includes($text as string)) {
				fireEvent.pointerDown($child, { pointerType: 'mouse' });
			}
		});

		return getChecked();
	};

	const pressRandomCheckbox = () => {
		const randomCheckbox = $children[faker.number.int($children.length - 1)];

		fireEvent.pointerDown(randomCheckbox, { pointerType: 'mouse' });

		return getChecked();
	};

	const pressRandomCheckboxes = (
		num = faker.number.int($children.length - 1),
	) => {
		const checkboxesToSelect = faker.helpers.shuffle($children).slice(0, num);

		checkboxesToSelect.forEach(($checkbox) => {
			fireEvent.pointerDown($checkbox, { pointerType: 'mouse' });
		});

		return getChecked();
	};

	return {
		$boxes,
		$checkmarks,
		$children,
		$component,
		$containers,
		$dashmarks,
		$fills,
		$inputs,
		$label,
		$labels,
		$root: $component,
		$strikethroughs,
		$text,
		checked,
		childCount,
		getChecked,
		getStrikethroughs,
		getValue,
		result,
		states,
		unchecked,
		value,
		values,
		pressCheckbox,
		pressCheckboxes,
		pressRandomCheckbox,
		pressRandomCheckboxes,
	};
};
