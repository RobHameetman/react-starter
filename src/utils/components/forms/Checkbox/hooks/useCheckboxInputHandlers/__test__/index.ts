export * from './changeEventWith';
export * from './context';
export * from './props';
export * from './randomState';
export * from './types';
