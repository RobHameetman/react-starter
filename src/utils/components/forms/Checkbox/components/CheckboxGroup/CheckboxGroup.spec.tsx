import { act } from '@testing-library/react';
import { matchMediaSpy } from '@test/spies/misc/matchMediaSpy';
import { onTest } from '@test/utils/onTest';
import { Intent } from '@/theme/enums/Intents';
import { randomIntent } from '@/theme/enums/Intents/__test__/randomIntent';
import { Size } from '@/theme/enums/Sizes';
import { randomSize } from '@/theme/enums/Sizes/__test__/randomSize';
import { CheckboxGroup } from './CheckboxGroup';
import { Checkbox } from '../../Checkbox';
import { anyCheckbox, render } from './__test__';

describe('<CheckboxGroup />', () => {
	let $boxes: ReadonlyArray<HTMLDivElement> | null = null;
	let $checkmarks: ReadonlyArray<HTMLDivElement> | null = null;
	let $children: ReadonlyArray<HTMLDivElement> | null = null;
	let $component: HTMLElement | null = null;
	let $containers: ReadonlyArray<HTMLDivElement> | null = null;
	let $dashmarks: ReadonlyArray<HTMLDivElement> | null = null;
	let $input: ReadonlyArray<HTMLInputElement> | null = null;
	let $label: HTMLLegendElement | null = null;
	let $labels: ReadonlyArray<HTMLLabelElement> | null = null;
	let $strikethroughs: ReadonlyArray<HTMLDivElement> | null = null;
	let checked: number | null = null;
	let childCount: number | null = null;
	let error: Error | null = null;
	let mockOnChange: jest.Mock | null = null;
	let mockOnChangeCapture: jest.Mock | null = null;
	let intent: Intent | null = null;
	let result: unknown = null;
	let size: Size | null = null;
	let states: ReadonlyArray<boolean> | null = null;
	let unchecked: number | null = null;
	let value: string | null = null;
	let values: ReadonlyArray<string> | null = null;
	let index = 1;

	beforeEach(() => {
		try {
			matchMediaSpy();

			onTest(index, {
				/**
				 * should be polymorphic
				 */
				2: () => {
					({ $component } = render(
						<CheckboxGroup as="section">
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));
				},
				/**
				 * should disable animations by default when the user prefers reduced motion
				 */
				3: () => {
					({ $component } = render(
						<CheckboxGroup>
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));
				},
				/**
				 * should allow for animations to be disabled explicitly
				 */
				4: () => {
					({ $component } = render(
						<CheckboxGroup animated={false}>
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));
				},
				/**
				 * should check all checkboxes initially when the state is externally set to true
				 */
				5: () => {
					({ $checkmarks, checked, childCount, unchecked } = render(
						<CheckboxGroup checked>
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));
				},
				/**
				 * should not check all checkboxes initially when the state is not externally set to true
				 */
				6: () => {
					({ $checkmarks, checked, childCount, unchecked } = render(
						<CheckboxGroup>
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));
				},
				/**
				 * should be stylable
				 */
				7: () => {
					({ $component } = render(
						<CheckboxGroup className="test">
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));
				},
				/**
				 * should use an overridable delimiter for group values
				 */
				8: () => {
					const { getValue, pressRandomCheckboxes } = render(
						<CheckboxGroup delimiter="|">
							<Checkbox>Tokyo, Japan</Checkbox>
							<Checkbox>Chicago, IL, United States</Checkbox>
							<Checkbox>Sydney, Australia</Checkbox>
						</CheckboxGroup>,
					);

					act(() => {
						checked = pressRandomCheckboxes(2);
						value = getValue();
					});
				},
				/**
				 * should be disablable
				 */
				9: () => {
					let pressRandomCheckbox: () => void;

					mockOnChange = jest.fn(() => {
						result = false;
					});

					({ $component, pressRandomCheckbox } = render(
						<CheckboxGroup onChange={mockOnChange} disabled>
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));

					act(() => {
						pressRandomCheckbox();
					});
				},
				/**
				 * should be identifiable
				 */
				10: () => {
					({ $component } = render(
						<CheckboxGroup id="test">
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));
				},
				/**
				 * should apply interaction intents to all children when an intent is defined explicitly
				 */
				11: () => {
					intent = randomIntent() as Intent;

					({ $children } = render(
						<CheckboxGroup intent={intent}>
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));

					result = $children
						.map(({ classList }) => classList)
						.filter((classList) =>
							Array.from(classList).some((className) =>
								className.includes(intent as Intent),
							),
						);
				},
				/**
				 * should be namable
				 */
				12: () => {
					({ $component } = render(
						<CheckboxGroup name="test">
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));
				},
				/**
				 * should use the ID as a name by default
				 */
				13: () => {
					({ $component } = render(
						<CheckboxGroup id="test">
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));
				},
				/**
				 * should accept a readonly state
				 */
				14: () => {
					let pressRandomCheckbox: () => void;

					mockOnChange = jest.fn(() => {
						result = false;
					});

					({ $component, pressRandomCheckbox } = render(
						<CheckboxGroup onChange={mockOnChange} readonly>
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));

					act(() => {
						pressRandomCheckbox();
					});
				},
				/**
				 * should apply standardized sizes to all children when a size is defined explicitly
				 */
				15: () => {
					size = randomSize() as Size;

					({ $children } = render(
						<CheckboxGroup size={size}>
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));

					result = $children
						.map(({ classList }) => classList)
						.filter((classList) =>
							Array.from(classList).some((className) =>
								className.includes(size as Size),
							),
						);
				},
				/**
				 * should strikethrough all checkbox labels initially when all checkboxes are checked
				 */
				16: () => {
					({ childCount, $strikethroughs } = render(
						<CheckboxGroup strikethrough checked>
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));
				},
				/**
				 * should not strikethrough all checkbox labels initially when all checkboxes are not checked
				 */
				17: () => {
					({ childCount, $strikethroughs } = render(
						<CheckboxGroup strikethrough>
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));
				},
				/**
				 * should allow the value to be controlled externally
				 */
				18: () => {
					({ $children, $checkmarks, checked, value } = render(
						<CheckboxGroup value={['Test 1', 'Test 2']}>
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));
				},
				/**
				 * should add a checkbox to the value when an unchecked checkbox is pressed
				 */
				19: () => {
					const { getValue, pressRandomCheckbox } = render(
						<CheckboxGroup>
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					);

					act(() => {
						checked = pressRandomCheckbox();
						value = getValue();
					});
				},
				/**
				 * should remove a checkbox from the value when a checked checkbox is pressed
				 */
				20: () => {
					const { getValue, pressCheckbox } = render(
						<CheckboxGroup value={['Test 1']}>
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					);

					act(() => {
						checked = pressCheckbox('Test 1');
						value = getValue();
					});
				},
				/**
				 * should position elements vertically when the "vertical" prop is `true`
				 */
				21: () => {
					({ $component } = render(
						<CheckboxGroup vertical>
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));
				},
				/**
				 * should handle change events correctly when all checkboxes are initially unchecked
				 */
				22: () => {
					let pressRandomCheckbox: () => number;

					mockOnChange = jest.fn((e) => {
						result = false;
					});

					mockOnChangeCapture = jest.fn();

					({ $component, pressRandomCheckbox } = render(
						<CheckboxGroup
							onChange={mockOnChange}
							onChangeCapture={mockOnChangeCapture}
						>
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));

					act(() => {
						pressRandomCheckbox();
					});
				},
				/**
				 * should handle change events correctly when all checkboxes are initially checked
				 */
				23: () => {
					let pressRandomCheckbox: () => number;

					mockOnChange = jest.fn((e) => {
						result = false;
					});

					mockOnChangeCapture = jest.fn();

					({ $component, pressRandomCheckbox } = render(
						<CheckboxGroup
							onChange={mockOnChange}
							onChangeCapture={mockOnChangeCapture}
							checked
						>
							<Checkbox>Test 1</Checkbox>
							<Checkbox>Test 2</Checkbox>
							<Checkbox>Test 3</Checkbox>
						</CheckboxGroup>,
					));

					act(() => {
						pressRandomCheckbox();
					});
				},
			});
		} catch (thrown) {
			error = !(thrown instanceof Error) ? (thrown as Error) : new Error();
			console.error(thrown);
		}
	});

	afterEach(() => {
		jest.restoreAllMocks();

		$boxes = null;
		$checkmarks = null;
		$children = null;
		$component = null;
		$containers = null;
		$dashmarks = null;
		$input = null;
		$label = null;
		$labels = null;
		$strikethroughs = null;
		checked = null;
		childCount = null;
		error = null;
		intent = null;
		mockOnChange = null;
		mockOnChangeCapture = null;
		result = null;
		size = null;
		states = null;
		unchecked = null;
		value = null;
		values = null;

		index++;
	});

	it('should render without props or children', () => {
		expect(index).toBe(1);
		expect(error).toBeNull();

		expect(() => render(<CheckboxGroup />)).not.toThrowError();
		expect(render(<CheckboxGroup />, 1).result).toMatchSnapshot();
	});

	it('should be polymorphic', () => {
		expect(index).toBe(2);
		expect(error).toBeNull();

		expect($component?.tagName).toBe('SECTION');
	});

	it('should disable animations by default when the user prefers reduced motion', () => {
		expect(index).toBe(3);
		expect(error).toBeNull();

		expect(window.matchMedia).toBeCalledWith('prefers-reduced-motion: reduce');
	});

	it('should allow for animations to be disabled explicitly', () => {
		expect(index).toBe(4);
		expect(error).toBeNull();

		expect($component).toMatchSnapshot();
		expect(window.matchMedia).not.toBeCalled();
	});

	it('should check all checkboxes initially when the state is externally set to true', () => {
		expect(index).toBe(5);
		expect(error).toBeNull();

		expect(checked).toBe(Number(childCount));
		expect(unchecked).toBe(0);
	});

	it('should not check all checkboxes initially when the state is not externally set to true', () => {
		expect(index).toBe(6);
		expect(error).toBeNull();

		expect(checked).toBe(0);
		expect(unchecked).toBe(Number(childCount));
	});

	it('should be stylable', () => {
		expect(index).toBe(7);
		expect(error).toBeNull();

		expect($component).not.toBeNull();
		expect($component).toHaveClass('test');
	});

	it('should use an overridable delimiter for group values', () => {
		expect(index).toBe(8);
		expect(error).toBeNull();

		expect(checked).toBe(2);
		expect(value).toStrictEqual(expect.stringContaining('|'));
	});

	it('should be disablable', () => {
		expect(index).toBe(9);
		expect(error).toBeNull();

		expect(result).toBeNull();
		expect(mockOnChange).not.toBeCalled();
	});

	it('should be identifiable', () => {
		expect(index).toBe(10);
		expect(error).toBeNull();

		expect($component).not.toBeNull();
		expect($component).toHaveAttribute('id', 'test');
	});

	it('should apply interaction intents to all children when an intent is defined explicitly', () => {
		expect(index).toBe(11);
		expect(error).toBeNull();

		expect(result).toHaveLength(($children ?? []).length);
	});

	it('should be nameable', () => {
		expect(index).toBe(12);
		expect(error).toBeNull();

		expect($component).toHaveAttribute('name', 'test');
	});

	it('should use the ID as a name by default', () => {
		expect(index).toBe(13);
		expect(error).toBeNull();

		expect($component).toHaveAttribute('name', 'test');
	});

	it('should accept a readonly state', () => {
		expect(index).toBe(14);
		expect(error).toBeNull();

		expect(result).toBeNull();
		expect(mockOnChange).not.toBeCalled();
	});

	it('should apply standardized sizes to all children when a size is defined explicitly', () => {
		expect(index).toBe(15);
		expect(error).toBeNull();

		expect(result).toHaveLength(($children ?? []).length);
	});

	it('should strikethrough all checkbox labels initially when all checkboxes are checked', () => {
		expect(index).toBe(16);
		expect(error).toBeNull();

		expect($strikethroughs).toHaveLength(Number(childCount));
	});

	it('should not strikethrough all checkbox labels initially when all checkboxes are not checked', () => {
		expect(index).toBe(17);
		expect(error).toBeNull();

		expect($strikethroughs).toHaveLength(0);
	});

	it('should allow the value to be controlled externally', () => {
		expect(index).toBe(18);
		expect(error).toBeNull();

		expect(value).toBe('Test 1,Test 2');
		expect(checked).toBe(2);
	});

	it('should add a checkbox to the value when an unchecked checkbox is pressed', () => {
		expect(index).toBe(19);
		expect(error).toBeNull();

		expect(checked).toBe(1);
		expect(value).toStrictEqual(anyCheckbox);
	});

	it('should remove a checkbox from the value when a checked checkbox is pressed', () => {
		expect(index).toBe(20);
		expect(error).toBeNull();

		expect(checked).toBe(0);
		expect(value).toBe('');
	});

	it('should position elements vertically when the "vertical" prop is `true`', () => {
		expect(index).toBe(21);
		expect(error).toBeNull();

		expect($component).toHaveClass('vertical');
	});

	it('should handle change events correctly when all checkboxes are initially unchecked', () => {
		expect(index).toBe(22);
		expect(error).toBeNull();

		expect(mockOnChange).toBeCalledTimes(1);
		expect(mockOnChangeCapture).toBeCalledTimes(1);
		expect(result).toBe(false);
	});

	it('should handle change events correctly when all checkboxes are initially checked', () => {
		expect(index).toBe(23);
		expect(error).toBeNull();

		expect(mockOnChange).toBeCalledTimes(1);
		expect(mockOnChangeCapture).toBeCalledTimes(1);
		expect(result).toBe(false);
	});
});
