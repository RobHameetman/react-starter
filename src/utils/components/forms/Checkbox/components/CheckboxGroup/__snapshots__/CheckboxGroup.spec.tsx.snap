// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxGroup /> should allow for animations to be disabled explicitly 1`] = `
<fieldset
  class="checkboxGroup checkboxGroup"
  id="checkboxGroup5"
  name="checkboxGroup5"
  role="group"
>
  <legend
    class="label md"
    id="checkboxGroup5-label"
  />
  <div
    aria-checked="false"
    aria-disabled="false"
    class="checkbox md none notAnimated"
    id="7"
    role="checkbox"
    tabindex="0"
  >
    <input
      id="7-input"
      name="7"
      role="hidden"
      tabindex="-1"
      title="7"
      type="checkbox"
      value=""
    />
    <div
      class="container"
      id="7-container"
    >
      <div
        class="box"
        id="7-box"
      />
      <div
        class="fill"
        id="7-fill"
      />
      <div
        class="check"
        id="7-checkmark"
      />
    </div>
    <label
      class="label"
      for="7-label"
    >
      Test 1
    </label>
  </div>
  <div
    aria-checked="false"
    aria-disabled="false"
    class="checkbox md none notAnimated"
    id="8"
    role="checkbox"
    tabindex="0"
  >
    <input
      id="8-input"
      name="8"
      role="hidden"
      tabindex="-1"
      title="8"
      type="checkbox"
      value=""
    />
    <div
      class="container"
      id="8-container"
    >
      <div
        class="box"
        id="8-box"
      />
      <div
        class="fill"
        id="8-fill"
      />
      <div
        class="check"
        id="8-checkmark"
      />
    </div>
    <label
      class="label"
      for="8-label"
    >
      Test 2
    </label>
  </div>
  <div
    aria-checked="false"
    aria-disabled="false"
    class="checkbox md none notAnimated"
    id="9"
    role="checkbox"
    tabindex="0"
  >
    <input
      id="9-input"
      name="9"
      role="hidden"
      tabindex="-1"
      title="9"
      type="checkbox"
      value=""
    />
    <div
      class="container"
      id="9-container"
    >
      <div
        class="box"
        id="9-box"
      />
      <div
        class="fill"
        id="9-fill"
      />
      <div
        class="check"
        id="9-checkmark"
      />
    </div>
    <label
      class="label"
      for="9-label"
    >
      Test 3
    </label>
  </div>
</fieldset>
`;

exports[`<CheckboxGroup /> should render without props or children 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <fieldset
        class="checkboxGroup checkboxGroup"
        id="checkboxGroup1"
        name="checkboxGroup1"
        role="group"
      >
        <legend
          class="label md"
          id="checkboxGroup1-label"
        />
      </fieldset>
    </div>
    <div>
      <fieldset
        class="checkboxGroup checkboxGroup"
        id="checkboxGroup2"
        name="checkboxGroup2"
        role="group"
      >
        <legend
          class="label md"
          id="checkboxGroup2-label"
        />
      </fieldset>
    </div>
  </body>,
  "container": <div>
    <fieldset
      class="checkboxGroup checkboxGroup"
      id="checkboxGroup2"
      name="checkboxGroup2"
      role="group"
    >
      <legend
        class="label md"
        id="checkboxGroup2-label"
      />
    </fieldset>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
